### ISSUE TITLE
[KAFKA] Add support to override consumer position and reset to head/tail

### ISSUE THREAD

**Issue by stanlemon** [salesforce/storm-dynamic-spout#29]

**Comment by Crim** [comment]
So wondering the best way to do this, as our startup logic is already fairly complicated.

Here's the 3 scenarios that exist today:
1. Consumer started with no explicit start position, and has no previous consumer state
2. Consumer started with no explicit start position, and has previous consumer state
3. Consumer started with explicit start position, and has no previous consumer state
4. Consumer started with explicit start position, and has previous consumer state

**Comment by Crim** [comment]
Yeah I guess the trick is to figure out how to only instruct the fire hose.  I think you'd need to use the virtual spout id.

What if we did the following:

Configuration key is something like: 

```yml
consumer.reset.default: HEAD|TAIL
consumer.reset.<virtualSpoutId>: HEAD|TAIL
```

Today we default to HEAD for reset.  If an offset is no good, we back up to head.  the default key makes this logic configurable.

the consumer.reset.default would work as follows:
```java
if (has no persisted state) {
  if (consumer.reset.default == HEAD) {
    if (hasStartingState) => reset to starting state;
    if (has No startingState) => reset to HEAD of topic;
  } else if (consumer.reset.default == tail) {
    if (hasEndingState) => reset to ending state;
    if (has NO endingState) => reset to tail of topic;
  }
} else if (has persisted state && consumer.reset.<my consumerid> != null) {
  if (consumer.reset.<myconsumerid> == HEAD) {
    if (hasStartingState) => reset to starting state;
    if (has No startingState) => reset to HEAD of topic;
  } else if (consumer.reset.<myconsumerid> == tail) {
    if (hasEndingState) => reset to ending state;
    if (has NO endingState) => reset to tail of topic;
  }
}
```

(Sweetjesus) thats complicated.

**Comment by stanlemon** [comment]
@Crim Consider this...
- The firehose in sidelining always starts with a null for starting. This is the easiest scenario to 'reset' on. If the reset option is set, the consumer just seeks to that point and moves on.
- Sidelines always start with an explicit starting point. In this situation reset to head is probably reset to the starting point.
- If a sideline is seeked past it's ending state (reset to tail), which is not stored in the consumer, the VirtualSpout will see that the consumer's offset is now past the ending offset of the vspout (sideline) and complete the vspout. See https://github.com/salesforce/storm-dynamic-spout/blob/master/src/main/java/com/salesforce/storm/spout/dynamic/VirtualSpout.java#L645-L693

So I think this actually could be relatively straight forward and Kafka-specific. I really think we add this reset position config with two options, check them in the Kafka consumer's open() call, do a seek and replace state accordingly and then let the rest of the spout work as intended, with this one caveat here https://github.com/salesforce/storm-dynamic-spout/blob/master/src/main/java/com/salesforce/storm/spout/dynamic/kafka/Consumer.java#L234-L236 where we think about what the 'head' of the consumer is a little differently.

If I get time today I may try to carve out what I'm talking about so you can look at it more concretely.
